In [1]:
!pip install datasets

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 480.6/480.6 kB 12.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 9.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 7.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 9.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 11.2 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2024.10.0
    Uninstalling fsspec-2024.10.0:
      Successfully uninstalled fsspec-2024.10.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2024.10.0 requires fsspec==2024.10.0, but you have fsspec 2024.9.0 which is incompatible.


## load dataset

In [2]:
from datasets import load_dataset

# Load the Yelp Review Full dataset
dataset = load_dataset("yelp_review_full", split = "all")


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/6.72k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/299M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/23.5M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/650000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/50000 [00:00<?, ? examples/s]

In [3]:
dataset

Dataset({
    features: ['label', 'text'],
    num_rows: 700000
})

In [4]:
# total_word = 0
# for i in range(len(dataset)):
#   total_word += len(dataset[i]['text'].split())
# print(total_word)
# # dataset[0]['text'].split()

In [5]:
93878307/len(dataset)

134.11186714285714

In [6]:
# total_single_word_line = 0
# for i in range(len(dataset)):
#   if len(dataset[i]['text'].split()) == 1:
#     total_single_word_line += 1
# total_single_word_line

In [7]:
"this is an apple"

'this is an apple'

In [8]:
# for i in range(len(dataset)):
#   if "Cheesy-melty-roasted-cauliflower-with-fresh-bread-crumbs-on top.\\nTo-die-for" in dataset[i]['text'].split():
#     print("yes")

## Check if **the distribution of the length of the samples is right skewed**

In [9]:
# word_len = list()
# for i in range(len(dataset)):
#   word_len.append(len(dataset[i]['text'].split(' ')))

In [10]:
# import matplotlib.pyplot as plt
# import numpy as np
# import random
# sorted_len = sorted(word_len)
# bins = np.unique(sorted_len)
# plt.figure(figsize=(10,4))
# plt.hist(sorted_len,bins=bins)
# plt.xlabel('Sequence length')
# plt.ylabel('count')
# plt.grid()
# plt.show()


In [11]:
from transformers import AutoTokenizer

# Load the pre-trained BERT tokenizer using AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


In [13]:
tokenizer('hello world')

{'input_ids': [101, 7592, 2088, 102], 'token_type_ids': [0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1]}

In [14]:
tokenizer

BertTokenizerFast(name_or_path='bert-base-uncased', vocab_size=30522, model_max_length=512, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("[PAD]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	100: AddedToken("[UNK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	101: AddedToken("[CLS]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	102: AddedToken("[SEP]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	103: AddedToken("[MASK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}

In [15]:
print("Vocabulary Size:", tokenizer.vocab_size)
print("Do Lower Case:", tokenizer.do_lower_case)
print("Padding Token:", tokenizer.pad_token)
print("UNK Token:", tokenizer.unk_token)
print("special tokens:", tokenizer.special_tokens_map)
print("context length:", tokenizer.model_max_length)


Vocabulary Size: 30522
Do Lower Case: True
Padding Token: [PAD]
UNK Token: [UNK]
special tokens: {'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}
context length: 512


In [16]:
from transformers import BertConfig, BertForMaskedLM

# Load the default BERT configuration
config = BertConfig()

# Initialize the BERT model for masked language modeling
model = BertForMaskedLM(config)

# Print model architecture and configuration
print(model)
print("Model configuration:")
print(config)

BertForMaskedLM(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSdpaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwi

In [17]:
# Count the total number of parameters
total_params = sum(p.numel() for p in model.parameters())

# Count the number of trainable parameters
trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)

print(f"Total parameters: {total_params}")
print(f"Trainable parameters: {trainable_params}")

Total parameters: 109514298
Trainable parameters: 109514298


## Note:

model.parameters(): This is a method provided by PyTorch models, which returns an iterator over all the parameters (learnable weights) of the model, including the weights and biases of layers like linear layers, embeddings, etc.

p.numel(): For each parameter p, the method .numel() returns the total number of elements in that tensor. For example, if p is a matrix with a shape of (768, 3072), then p.numel() would return
768
×
3072
=2,359,296


In [18]:
# Step 1: Load the default config and modify max_position_embeddings
config = BertConfig.from_pretrained('bert-base-uncased')

# Modify the maximum sequence length (context length)
config.max_position_embeddings = 1024  # Change from 512 to 1024 tokens

# Step 2: Initialize the model with the modified config
model_ = BertForMaskedLM(config)

# Now the model has the updated context length
print(model_.config)


BertConfig {
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 1024,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "transformers_version": "4.44.2",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
}



In [19]:
sum(p.numel() for p in model_.parameters()) - trainable_params


393216

## Preprocess

In [20]:
import numpy as np
from datasets import Dataset
# Define a function to tokenize and chunk the data
def chunk_samples(batch):
    # Tokenize the batch, getting input IDs and attention masks
    tokenized_batch = tokenizer(batch) #, padding=False, truncation=False, return_tensors="np")

    # Concatenate all input IDs
    # input_ids_concat = np.concatenate(tokenized_batch['input_ids'])
    input_ids_concat = sum(tokenized_batch['input_ids'], [])

    # Chunk the concatenated input IDs into size of 512
    num_chunks = len(input_ids_concat) // 512
    chunks = [input_ids_concat[i*512:(i+1)*512] for i in range(num_chunks)]

    # Return the packed chunks
    return chunks

# Initialize an empty list to hold all the chunks
all_chunks = []

# Process the dataset in batches of 1000
batch_size = 1000
for i in range(0, len(dataset), batch_size):
    # Extract the batch text
    batch_text = dataset['text'][i:i+batch_size]

    # Tokenize and chunk the batch
    chunks = chunk_samples(batch_text)

    # Add the chunks to the list
    all_chunks.extend(chunks)

# Create a new dataset from the chunks
ds_chunked = Dataset.from_dict({'input_ids': all_chunks})

# Get the total number of new samples (chunks) in the resulting dataset
total_samples = len(ds_chunked['input_ids'])

print(f"Total number of samples (chunks) in the new dataset: {total_samples}")

Token indices sequence length is longer than the specified maximum sequence length for this model (1019 > 512). Running this sequence through the model will result in indexing errors


Total number of samples (chunks) in the new dataset: 246695


In [21]:
all_chunks[0]

[101,
 2852,
 1012,
 18522,
 4107,
 2673,
 1045,
 2298,
 2005,
 1999,
 1037,
 2236,
 18742,
 1012,
 2002,
 1005,
 1055,
 3835,
 1998,
 3733,
 2000,
 2831,
 2000,
 2302,
 2108,
 9161,
 6026,
 1025,
 2002,
 1005,
 1055,
 2467,
 2006,
 2051,
 1999,
 3773,
 2010,
 5022,
 1025,
 2002,
 1005,
 1055,
 6989,
 2007,
 1037,
 2327,
 1011,
 18624,
 2902,
 1006,
 27935,
 1007,
 2029,
 2026,
 3008,
 2031,
 4541,
 2000,
 2033,
 2003,
 2200,
 2590,
 1999,
 2553,
 2242,
 6433,
 1998,
 2017,
 2342,
 5970,
 1025,
 1998,
 2017,
 2064,
 2131,
 6523,
 7941,
 2015,
 2000,
 2156,
 15744,
 2302,
 2383,
 2000,
 2156,
 2032,
 2034,
 1012,
 2428,
 1010,
 2054,
 2062,
 2079,
 2017,
 2342,
 1029,
 1045,
 1005,
 1049,
 3564,
 2182,
 2667,
 2000,
 2228,
 1997,
 2151,
 10821,
 1045,
 2031,
 2055,
 2032,
 1010,
 2021,
 1045,
 1005,
 1049,
 2428,
 5059,
 1037,
 8744,
 1012,
 102,
 101,
 6854,
 1010,
 1996,
 9135,
 1997,
 2108,
 2852,
 1012,
 18522,
 1005,
 1055,
 5776,
 2003,
 1037,
 9377,
 1997,
 1996,
 3325,
 1045,
 1

In [22]:
tokenizer('[MASK]')

{'input_ids': [101, 103, 102], 'token_type_ids': [0, 0, 0], 'attention_mask': [1, 1, 1]}

In [23]:
tokenizer("lets go to extreme", padding=False, truncation=True, return_tensors="np")


{'input_ids': array([[  101, 11082,  2175,  2000,  6034,   102]]), 'token_type_ids': array([[0, 0, 0, 0, 0, 0]]), 'attention_mask': array([[1, 1, 1, 1, 1, 1]])}

In [24]:
from transformers import DataCollatorForLanguageModeling
from torch.utils.data import DataLoader
from sklearn.model_selection import train_test_split
from datasets import Dataset, DatasetDict

# Split the dataset into training and test sets (5% test size)
train_dataset, test_dataset = ds_chunked.train_test_split(test_size=0.05, seed=42).values()

# Define the MLM data collator with a masking probability of 0.2
data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer,
    mlm=True,
    mlm_probability=0.2
)

# Convert the Hugging Face datasets into PyTorch tensors
train_dataset.set_format(type='torch', columns=['input_ids'])
test_dataset.set_format(type='torch', columns=['input_ids'])

# Use PyTorch DataLoader to load batches of data
train_dataloader = DataLoader(train_dataset, batch_size=8, shuffle=True, collate_fn=data_collator)
test_dataloader = DataLoader(test_dataset, batch_size=8, shuffle=False, collate_fn=data_collator)

# Fetch a batch of data from the train dataloader
batch = next(iter(train_dataloader))

# The input_ids after masking (collator applies masking)
masked_input_ids = batch['input_ids']

# Token IDs corresponding to unmasked tokens can be found as follows:
# If you want to inspect unmasked tokens (which remain unchanged), compare with original input
# In BERT, usually [MASK] token ID is 103, so you can find unmasked tokens where they are not equal to 103.
mask_token_id = tokenizer.mask_token_id
unmasked_tokens = masked_input_ids[masked_input_ids != mask_token_id]

# print("Masked Input IDs:\n", masked_input_ids)
# print("Unmasked Tokens:\n", unmasked_tokens)

# # If you want to get the token IDs corresponding to the unmasked tokens
# print(f"Unmasked token IDs: {unmasked_tokens.tolist()}")

In [25]:
len(unmasked_tokens), len(masked_input_ids[0])

(3446, 512)

In [26]:
config = BertConfig(
    num_hidden_layers=6,
    hidden_size=384,
    intermediate_size=1536
)
# Step: Initialize the BERT model for MLM (Masked Language Modeling)
model = BertForMaskedLM(config)

In [35]:
train_dataset = train_dataset.select(range(10000))
test_dataset = test_dataset.select(range(1000))


In [36]:
from transformers import Trainer, TrainingArguments

# Define the training arguments
training_args = TrainingArguments(
    output_dir='./results',          # Output directory
    overwrite_output_dir=True,       # Overwrite the content of the output directory
    num_train_epochs=3,              # Number of training epochs
    per_device_train_batch_size=8,   # Batch size per device during training
    per_device_eval_batch_size=8,   # Batch size per device during evaluation
)

trainer = Trainer(
    model=model,
    args=training_args,
    data_collator=data_collator,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,

)


In [37]:
trainer.train()

TrainOutput(global_step=3750, training_loss=6.22223564453125, metrics={'train_runtime': 966.6837, 'train_samples_per_second': 31.034, 'train_steps_per_second': 3.879, 'total_flos': 997787013120000.0, 'train_loss': 6.22223564453125, 'epoch': 3.0})

In [38]:
model


BertForMaskedLM(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 384, padding_idx=0)
      (position_embeddings): Embedding(512, 384)
      (token_type_embeddings): Embedding(2, 384)
      (LayerNorm): LayerNorm((384,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-5): 6 x BertLayer(
          (attention): BertAttention(
            (self): BertSdpaSelfAttention(
              (query): Linear(in_features=384, out_features=384, bias=True)
              (key): Linear(in_features=384, out_features=384, bias=True)
              (value): Linear(in_features=384, out_features=384, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=384, out_features=384, bias=True)
              (LayerNorm): LayerNorm((384,), eps=1e-12, elementwise